service/session_http.go

```diff
@@ -0,0 +1,160 @@
+	"joshsoftware/go-e-commerce/config"
+	"joshsoftware/go-e-commerce/db"
+	"net/http"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	logger "github.com/sirupsen/logrus"
+)
+
+//AuthBody stores responce body for login
+type AuthBody struct {
+	Message string `json:"meassage"`
+	Token   string `json:"token"`
+}
+
+//generateJWT function generates and return a new JWT token
+func generateJwt(userID int) (tokenString string, err error) {
+	mySigningKey := config.JWTKey()
+	if mySigningKey == nil {
+		ae.Error(ae.ErrNoSigningKey, "Application error: No signing key configured", err)
+		return
+	}
+
+	token := jwt.New(jwt.SigningMethodHS256)
+	claims := token.Claims.(jwt.MapClaims)
+	claims["id"] = userID
+	claims["exp"] = time.Now().Add(time.Duration(config.JWTExpiryDurationHours()) * time.Hour).Unix()
+
+	tokenString, err = token.SignedString(mySigningKey)
+	if err != nil {
+		ae.Error(ae.ErrSignedString, "Failed To Get Signed String", err)
+		return
+	}
+	return
+}
+
+//userLoginHandler function take credentials in json
+// and check if the credentials are correct
+// also generate and returns a JWT token in the case of correct crendential
+func userLoginHandler(deps Dependencies) http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		user := db.User{}
+
+		//fetching the json object to get crendentials of users
+		err := json.NewDecoder(req.Body).Decode(&user)
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("JSON Decoding Failed")
+			rw.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		//TODO change no need to return user object from Authentication
+		//checking if the user is authenticated or not
+		// by passing the credentials to the AuthenticateUser function
+		user, err1 := deps.Store.AuthenticateUser(req.Context(), user)
+		if err1 != nil {
+			logger.WithField("err", err1.Error()).Error("Invalid Credentials")
+			rw.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+
+		//Generate new JWT token if the user is authenticated
+		// and return the token in request header
+		token, err := generateJwt(user.ID)
+		if err != nil {
+			rw.WriteHeader(http.StatusInternalServerError)
+			rw.Write([]byte("Token Generation Failure"))
+			return
+		}
+		authbody := AuthBody{
+			Message: "Login Successfull",
+			Token:   token,
+		}
+
+		respBytes, err := json.Marshal(authbody)
+		if err != nil {
+			ae.Error(ae.ErrJSONParseFail, "JSON Parsing Failed", err)
+		}
+
+		rw.Header().Add("Content-Type", "application/json")
+		rw.Write(respBytes)
+	})
+}
+
+//userLogoutHandler function logs the user off
+// and add the valid JWT token in BlacklistedToken
+func userLogoutHandler(deps Dependencies) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+
+		//fetching the token from header
+		authToken := req.Header["Token"]
+
+		//fetching details from the token
+		userID, expirationTimeStamp, err := getDataFromToken(authToken[0])
```

Comment:
Here, `authToken[0]` will be index out of range if we don't pass the `Token` in header.. can we fix this?

or we can use `Header.Get` method

Reply:
We will use Header.Get method